### PR TITLE
feat: adding cache scavenger

### PR DESCRIPTION
**Pull request by tedxu**
See #31262

### CODE REVIEW

**Comment by tedxu** [comment]
rerun ut

File: pkg/util/cache/cache.go [hunk 1]
```diff
@@ -29,15 +35,55 @@ func (i *cacheItem[K, V]) Value() V {
...
+func NewLazyScavenger[K comparable](weight func(K) int64, capacity int64) *LazyScavenger[K] {
+	return &LazyScavenger[K]{
+		capacity: capacity,
+		weight:   weight,
+	}
+}
+
+func (s *LazyScavenger[K]) Collect(key K) (bool, func(K) bool) {
+	w := s.weight(key)
+	if s.size+w > s.capacity {
```
Comment (sunby) [inline pkg/util/cache/cache.go:70]: maybe have data race if multiple threads access and modify size here.

File: pkg/util/cache/cache.go [hunk 2]
```diff
@@ -81,100 +187,94 @@ func (c *lruCache[K, V]) GetAndPin(key K) (*cacheItem[K, V], bool) {
...
-
-	_, ok := c.items[key]
-	return ok
+func (c *lruCache[K, V]) tryScavenge(key K) ([]K, bool) {
+	ok, collector := c.scavenger.Collect(key)
+	toEvict := make([]K, 0)
+	if !ok {
+		// Try evict to make room
+		done := false
+		for p := c.accessList.Back(); p != nil && !done; p = p.Prev() {
```
Comment (sunby) [inline pkg/util/cache/cache.go:219]: I am not sure whether `accessList` is thread(goroutine) safe.
If we have a goroutine that is pushing item to `accessList` and another is iterating it, maybe there's data race problem.

File: pkg/util/cache/cache.go [hunk 3]
```diff
@@ -47,30 +93,90 @@ type lruCache[K comparable, V any] struct {
...
+	item, err := c.getAndPin(key)
+	if err != nil {
+		return err
+	}
+	defer item.Unpin()
+
+	if err := doer(item.Value()); err != nil {
+		return err
 	}
+	return nil
```
Comment (sunby) [inline pkg/util/cache/cache.go:174]: small optimization: can just `return doer(item.Value())`

File: pkg/util/cache/cache_test.go [hunk 4]
```diff
@@ -3,36 +3,123 @@ package cache
...
+				return nil
+			})
+			assert.NoError(t, err)
+		}
+		assert.Equal(t, []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18}, finalizeSeq)
+		err := cache.Do(100, func(v int) error {
+			return nil
+		})
+		assert.Equal(t, ErrNotEnoughSpace, err)
+	})
```
Comment (sunby) [inline pkg/util/cache/cache_test.go:108]: we can add a test for multiple goroutine case to avoid data race

**Pull request merged**